### ISSUE TITLE
[MODULES-171] Cannot iterate over META-INF contents from imported modules

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#605]
@tdiesler opened **[MODULES-171](https://issues.redhat.com/browse/MODULES-171?redirect=false)** and commented

A module statically imports a few component modules like this

```
<module xmlns="urn:jboss:module:1.1" name="org.wildfly.camel">
    ...        

    <dependencies>
        ...
        <!-- components -->
        <module name="org.apache.camel.component.cxf" services="import"/>
        <module name="org.apache.camel.component.jms" services="import"/>
        <module name="org.apache.camel.component.jmx" services="import"/>
    </dependencies>

</module>
```

They all contain property files named after the component in

```
META-INF/services/org/apache/camel/component
```

All of these approaches to iterate over the resources don't work

```
//itres = module.iterateResources(PathFilters.getMetaInfFilter());
//itres = module.iterateResources(PathFilters.isChildOf("META-INF/services/org/apache/camel/component"));
//itres = module.iterateResources(PathFilters.isChildOf(DefaultComponentResolver.RESOURCE_PATH));
//itres = classLoader.iterateResources(DefaultComponentResolver.RESOURCE_PATH, true);
```

Instead, I need to use a hard coded list of the actual component modules and iterate over their respective local resources like this

```
// Remove the trailing slash from DefaultComponentResolver.RESOURCE_PATH
Iterator<Resource> itres = classLoader.iterateResources("META-INF/services/org/apache/camel/component", true);
```

---

**Affects:** 1.2.2.Final

**Comment by dmlloyd** [comment]
@dmlloyd commented

Have you tried 1.2.3.Final?  This includes an iteration fix.

**Comment by dmlloyd** [comment]
@tdiesler commented

Unfortunately, 1.2.3.Final does not make a difference.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Are you aware that `isChildOf` will exclude the named directory, only including files in directories that are children of the given directory?  You might want `any(isChildOf(path), is(path))` instead.

**Comment by dmlloyd** [comment]
@tdiesler commented

This is not the cause either. AFAICS, there is no test coverage for the Module.iterateResources(...) API. Are you sure this is working as you intend? Is it possible that services="import" does not support subdirectories of META-INF/services ?

**Comment by dmlloyd** [comment]
@tdiesler commented

Filed PR [#40](https://github.com/jboss-modules/jboss-modules/pull/40) that includes subdirectories from META-INF/services in PathFilters.getMetaInfServicesFilter()

The rational is that when you do

```
<dependencies>
    ...
    <!-- components -->
    <module name="org.apache.camel.component.cxf" services="import"/>
    <module name="org.apache.camel.component.jms" services="import"/>
    <module name="org.apache.camel.component.jmx" services="import"/>
</dependencies>
```

you also see META-INF/services/**

**Comment by dmlloyd** [comment]
@dmlloyd commented

We only have had manual, out of tree testing for iterateResources() on a module basis, but as far as I know it is working properly.

Service import works exactly as intended currently.  That said, the intent might be wrong.  It is designed to meet the ServiceLoader spec only (i.e. only files directly in the services directory are recognized) but if there are de facto implementations which go beyond this then it should be changed.  So pursuant to that, I do not have any problem with accepting PR#40 (once I review it in detail anyway).

**Comment by dmlloyd** [comment]
**[Brad Maxwell](https://issues.redhat.com/secure/ViewProfile.jspa?name=bmaxwell)** commented

It looks like this should be marked resolved in 1.2.4.Final, commit f1f32c1872ee4914d3a993dbf7ca088854c37d14 is merged.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Will do, thanks Brad.
